### PR TITLE
[Improvement] max_batches support to training log and tqdm progress bar.

### PR DESCRIPTION
**Pull request by hakuryuu96**
### Issue description
As @BloodAxe described, if _training_hyperparams.max_train/valid_batches_ are redefined in CLI, _tqdm_ and training log do not take changes into account and continue showing the full length of the dataloader. E.g. when user executes something like this

```
train_params = {
    "max_epochs": 300,
    "phase_callbacks": phase_callbacks,
    "initial_lr": lr,
    "loss": loss_fn,
    "optimizer": optimizer,
    "train_metrics_list": [Accuracy(), Top5()],
    "valid_metrics_list": [Accuracy(), Top5()],
    "metric_to_watch": "Accuracy",
    "greater_metric_to_watch_is_better": True,
    "lr_scheduler_step_type": "epoch",
    "max_train_batches": 24,
    "max_valid_batches": 24,
}

trainer.train(model=net, training_params=train_params, train_loader=train_loader, valid_loader=valid_loader)
```
the resulting logs and progress bar are the following:
```
[2023-10-19 13:30:41] INFO - sg_trainer_utils.py - TRAINING PARAMETERS:
    - Mode:                         Single GPU
    - Number of GPUs:               1          (2 available on the machine)
    - Full dataset size:            50000      (len(train_set))
    - Batch size per GPU:           16         (batch_size)
    - Batch Accumulate:             1          (batch_accumulate)
    - Total batch size:             16         (num_gpus * batch_size)
    - Effective Batch size:         16         (num_gpus * batch_size * batch_accumulate)
    - Iterations per epoch:         3125         (len(train_loader))
    - Gradient updates per epoch:   3125         (len(train_loader) / batch_accumulate)

[2023-10-19 13:30:41] WARNING - sg_trainer_utils.py - max_train_batch is set to 24. This limits the number of iterations per epoch and gradient updates per epoch.
[2023-10-19 13:30:41] INFO - sg_trainer.py - Started training for 300 epochs (0/299)

Train epoch 0: 1%|█                        | 24/3125 [00:02<00:00,  9.22it/s, Accuracy=0.0729, CrossEntropyLoss=2.53, Top5=0.508, gpu_mem=0.231]
Validating: 1%|█                      | 24/3125 [00:00<00:00, 126.12it/s]
2023-10-19 13:30:44] INFO - base_sg_logger.py - Checkpoint saved in /home/phil/deci/super-gradients/checkpoints/Cifar10_external_objects_example/RUN_20231019_133041_486426/ckpt_best.pth
[2023-10-19 13:30:44] INFO - sg_trainer.py - Best checkpoint overriden: validation Accuracy: 0.1002604141831398
===========================================================
SUMMARY OF EPOCH 0
├── Train
│   ├── Crossentropyloss = 2.5337
│   ├── Accuracy = 0.0729
│   └── Top5 = 0.5078
└── Validation
    ├── Crossentropyloss = 2.3486
    ├── Accuracy = 0.1003
    └── Top5 = 0.4596

===========================================================
```
### PR description
This PR addresses the issue above and proposes some improvements. Briefly:
1. Logs show the actual number of used elements in dataloader 
    1.5. Additionally, logs are warning user that the _max_batches_ parameter was set.
2. Progress bar shows the actual number of steps it will take to finish the epoch

E.g. if the _max_train/valid_batches_ parameter is specified:
```
[2023-10-19 13:30:41] INFO - sg_trainer_utils.py - TRAINING PARAMETERS:
    - Mode:                         Single GPU
    - Number of GPUs:               1          (2 available on the machine)
    - Full dataset size:            50000      (len(train_set))
    - Batch size per GPU:           16         (batch_size)
    - Batch Accumulate:             1          (batch_accumulate)
    - Total batch size:             16         (num_gpus * batch_size)
    - Effective Batch size:         16         (num_gpus * batch_size * batch_accumulate)
    - Iterations per epoch:         24         (len(train_loader) OR max_train_batches)
    - Gradient updates per epoch:   24         (len(train_loader) OR max_train_batches / batch_accumulate)

[2023-10-19 13:30:41] WARNING - sg_trainer_utils.py - max_train_batch is set to 24. This limits the number of iterations per epoch and gradient updates per epoch.
[2023-10-19 13:30:41] INFO - sg_trainer.py - Started training for 300 epochs (0/299)

Train epoch 0: 100%|██████████| 24/24 [00:02<00:00,  9.22it/s, Accuracy=0.0729, CrossEntropyLoss=2.53, Top5=0.508, gpu_mem=0.231]
Validating: 100%|██████████| 24/24 [00:00<00:00, 126.12it/s]
[2023-10-19 13:30:44] INFO - base_sg_logger.py - Checkpoint saved in /home/phil/deci/super-gradients/checkpoints/Cifar10_external_objects_example/RUN_20231019_133041_486426/ckpt_best.pth
[2023-10-19 13:30:44] INFO - sg_trainer.py - Best checkpoint overriden: validation Accuracy: 0.1002604141831398
```
If not, the logs behave similarly to previous versions.

### Some ideas
IMO it should be cool to consider logging the whole set of training parameters before the run. For me as a user, it would be nice to double-check all the settings I've made somewhere in the project (e.g. if I use _hydra_ and take SG Trainer class to my pipeline) and to be sure things go smoothly :) 
For example, the user should see the following info:
- dataset (number of classes, class names, etc)
- dataloader (batch_size, num_workers, etc)
- model (model name, number of trainable parameters, etc)
- optimization info (optimizer, lr, wd, losses, etc)
- training info (number of epochs, number of gradient updates, EMA, SyncBN usage, etc)

### CODE REVIEW

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 1]
```diff
@@ -471,6 +475,9 @@ def _train_epoch(self, context: PhaseContext, silent_mode: bool = False) -> tupl
             context.update_context(loss_avg_meter=loss_avg_meter, metrics_compute_fn=self.train_metrics)
 
             for batch_idx, batch_items in enumerate(progress_bar_train_loader):
+                if self.max_train_batches is not None and self.max_train_batches <= batch_idx:
```
Comment (BloodAxe) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:478]: Perhaps we can use `expected_iterations` here instead of checking again whether self.max_train_batches is not None. Think it would make code a little bit less repeatnig

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 2]
```diff
@@ -2091,9 +2108,11 @@ def evaluate(
                 if epoch:
                     pbar_start_msg += f" epoch {epoch}"
                 progress_bar_data_loader.set_description(pbar_start_msg)
-
             with torch.no_grad():
                 for batch_idx, batch_items in enumerate(progress_bar_data_loader):
+                    if evaluation_type == EvaluationType.VALIDATION and self.max_valid_batches is not None and self.max_valid_batches <= batch_idx:
```
Comment (BloodAxe) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:2113]: Perhaps should use function argument `max_batches` here instead?

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 1 above)
Comment (hakuryuu96) [reply src/super_gradients/training/sg_trainer/sg_trainer.py]: Good point, thx!

**Review by BloodAxe — APPROVED** [review]
LGTM

**Review by Louis-Dupont — APPROVED** [review]
LGTM
